### PR TITLE
Migrate pallet-grandpa to attribute macros

### PR DESCRIPTION
**Pull request by octol**
Part of #7882.

Migrate `pallet-grandpa` to attribute macros.
See note about storage prefix. This will need a storage migration, which is done in the polkadot companion https://github.com/paritytech/polkadot/pull/2994
 
Following the upgrade guidelines here: https://crates.parity.io/frame_support/attr.pallet.html#upgrade-guidelines

## ⚠️ Breaking Change ⚠️ 

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storages now use PalletInfo for module_prefix instead of the one given to decl_storage: Thus any use of this pallet in construct_runtime! should be careful to update name in order not to break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the `Grandpa` pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the `Grandpa` pallet.

---

polkadot companion: https://github.com/paritytech/polkadot/pull/2994

### CODE REVIEW

**Comment by octol** [comment]
Comparing the metadata before and after I've noticed that the prefix is explicitly set to `GrandpaFinality` in `decl_module`, while with these changes it's now `Grandpa`. It's used as `Grandpa` in `construct_runtime!` in polkadot so I think there it should be fine? From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storage items defined with pallet use the name of the pallet provided by traits::PalletInfo::name as pallet_prefix (in decl_storage, storage items used the pallet_prefix given as input of decl_storage with the syntax as Example). Thus a runtime using the pallet must be careful with this change. To handle this change:
> - either ensure that the name of the pallet given to construct_runtime! is the same as the name the pallet was giving to decl_storage,
> - or do a storage migration from the old prefix used to the new prefix used.
>
> NOTE: The prefixes used by storage items are in the metadata. Thus, ensuring the metadata hasn’t changed does ensure that the pallet_prefixs used by the storage items haven’t changed.

So even if polkadot isn't affected, other users of the pallet still need to be careful about this

**Comment by octol** [comment]
> Comparing the metadata before and after I've noticed that the prefix is explicitly set to `GrandpaFinality` in `decl_module`, while with these changes it's now `Grandpa`. It's used as `Grandpa` in `construct_runtime!` in polkadot so I think there it should be fine? From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines
> 
> > storage items defined with pallet use the name of the pallet provided by traits::PalletInfo::name as pallet_prefix (in decl_storage, storage items used the pallet_prefix given as input of decl_storage with the syntax as Example). Thus a runtime using the pallet must be careful with this change. To handle this change:
> > 
> > * either ensure that the name of the pallet given to construct_runtime! is the same as the name the pallet was giving to decl_storage,
> > * or do a storage migration from the old prefix used to the new prefix used.
> > 
> > NOTE: The prefixes used by storage items are in the metadata. Thus, ensuring the metadata hasn’t changed does ensure that the pallet_prefixs used by the storage items haven’t changed.
> 
> So even if polkadot isn't affected, other users of the pallet still need to be careful about this

Checking with @thiolliere seems we need to either rename to `GrandpaFinality` in Polkadot, or do a storage migration. I leaning towards the former, so I'll raise a Polkadot PR

**Comment by octol** [comment]
> > Comparing the metadata before and after I've noticed that the prefix is explicitly set to `GrandpaFinality` in `decl_module`, while with these changes it's now `Grandpa`. It's used as `Grandpa` in `construct_runtime!` in polkadot so I think there it should be fine? From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines
> > > storage items defined with pallet use the name of the pallet provided by traits::PalletInfo::name as pallet_prefix (in decl_storage, storage items used the pallet_prefix given as input of decl_storage with the syntax as Example). Thus a runtime using the pallet must be careful with this change. To handle this change:
> > > 
> > > * either ensure that the name of the pallet given to construct_runtime! is the same as the name the pallet was giving to decl_storage,
> > > * or do a storage migration from the old prefix used to the new prefix used.
> > > 
> > > NOTE: The prefixes used by storage items are in the metadata. Thus, ensuring the metadata hasn’t changed does ensure that the pallet_prefixs used by the storage items haven’t changed.
> > 
> > 
> > So even if polkadot isn't affected, other users of the pallet still need to be careful about this
> 
> Checking with @thiolliere seems we need to either rename to `GrandpaFinality` in Polkadot, or do a storage migration. I leaning towards the former, so I'll raise a Polkadot PR

In the end I went down the path of a migration

File: bin/node-template/runtime/Cargo.toml [hunk 1]
```diff
@@ -16,7 +16,7 @@ codec = { package = "parity-scale-codec", version = "2.0.0", default-features =
 pallet-aura = { version = "3.0.0", default-features = false, path = "../../../frame/aura" }
 pallet-balances = { version = "3.0.0", default-features = false, path = "../../../frame/balances" }
 frame-support = { version = "3.0.0", default-features = false, path = "../../../frame/support" }
-pallet-grandpa = { version = "3.0.0", default-features = false, path = "../../../frame/grandpa" }
+pallet-grandpa = { version = "4.0.0", default-features = false, path = "../../../frame/grandpa" }
```
Comment (octol) [inline bin/node-template/runtime/Cargo.toml:19]: I'm not entirely sure about this. Seems awfully convenient, but on the other hand we don't seem the change these unless we update the substrate version?

File: frame/grandpa/src/migrations/v4.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+/// Some checks prior to migration. This can be linked to
+/// [`frame_support::traits::OnRuntimeUpgrade::pre_upgrade`] for further testing.
+///
+/// Panics if anything goes wrong.
+pub fn pre_migration<P: GetPalletVersion, N: AsRef<str>>(new: N) {
+	let new = new.as_ref();
+	log::info!("pre-migration grandpa test with new = {}", new);
+
+	// the next key must exist, and start with the hash of `OLD_PREFIX`.
+	let next_key = sp_io::storage::next_key(OLD_PREFIX).unwrap();
```
Comment (thiolliere) [inline frame/grandpa/src/migrations/v4.rs:82]: here you need to do `twox_128(OLD_PREFIX)`

File: frame/grandpa/src/migrations/v4.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	assert!(next_key.starts_with(&sp_io::hashing::twox_128(OLD_PREFIX)));
+
+	// ensure nothing is stored in the new prefix.
+	assert!(
+		sp_io::storage::next_key(new.as_bytes()).map_or(
```
Comment (thiolliere) [inline frame/grandpa/src/migrations/v4.rs:87]: here you need to do `twox_128(new.as_bytes())`

File: frame/grandpa/Cargo.toml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "pallet-grandpa"
-version = "3.0.0"
+version = "4.0.0"
```
Comment (octol) [inline frame/grandpa/Cargo.toml:3]: Appears the Polkadot PR check fails with the version bumped

File: frame/grandpa/src/migrations/v4.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+			// either nothing is there
+			true,
+			// or we ensure that it has no common prefix with twox_128(new).
+			|next_key| !next_key.starts_with(&sp_io::hashing::twox_128(new.as_bytes()))
+		),
+		"unexpected next_key({}) = {:?}",
+		new,
+		sp_core::hexdisplay::HexDisplay::from(&sp_io::storage::next_key(new.as_bytes()).unwrap())
+	);
```
[the remark this answers is not in the document]
Comment (thiolliere) [reply frame/grandpa/src/migrations/v4.rs:109]: I still this comment is valid no ? the version should use the pallet name here making the assert fail if I'm correct

[the remark this answers is not in the document]
Comment (thiolliere) [reply frame/grandpa/src/migrations/v4.rs:109]: ah my bad it is indeed fixed

**Comment by octol** [comment]
Changed the version, bump to 3.1.0 instead of 4.0.0

**Review by andresilva — APPROVED** [review]
lgtm

**Comment by andresilva** [comment]
bot merge

**Pull request merged**